staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go

```diff
@@ -196,6 +196,48 @@ func (c *CloudNodeLifecycleController) MonitorNodes(ctx context.Context) {
 	}
 }
 
+// getProviderID returns the provider ID for the node. If Node CR has no provider ID,
+// it will be the one from the cloud provider.
+func getProviderID(ctx context.Context, cloud cloudprovider.Interface, node *v1.Node) (string, error) {
+	if instanceV2, ok := cloud.InstancesV2(); ok {
+		metadata, err := instanceV2.InstanceMetadata(ctx, node)
+		if err != nil {
+			if err == cloudprovider.InstanceNotFound {
+				return "", nil
+			}
+			return "", err
+		}
+
+		return metadata.ProviderID, nil
+	}
+
+	instances, ok := cloud.Instances()
+	if !ok {
+		return "", errors.New("cloud provider does not support instances")
+	}
```

Comment:
this code is called before in the two functions that call this one, why do we need to duplicate the cloud provider calls?

Reply:
I did see it when I wrote it but I had a trouble simply removing that call... The way I think is that I have `getProviderID` which takes `cloud cloudprovider.Interface` as one of parameters. Sure, it's true that `shutdownInCloudProvider` and `ensureNodeExistsByProviderID` call `cloud.InstancesV2()` already but what if someone introduces a 3rd function that uses `getProviderID` somewhere in its body but doesn't check for V2 before? 

Today, with only 2 consumers, we have guarantee that `getProviderID` is only called for non-V2. But tomorrow, how to ensure this is the case? If you have any suggestion I am more than happy to hear!

Reply:
this is a DRY vs WET example to me https://www.zippyops.com/is-your-code-dry-or-wet#:~:text=DRY%20code%20is%20a%20software,t%20adhere%20to%20DRY%20principle., just repeat the code twice, if there ia third function that wants to use it that is the moment to consolidate it

My 2 cents

Reply:
It is tempting to remove both `if instanceV2, ok := cloud.InstancesV2();` and `instances, ok := cloud.Instances()` and simply say that whoever calls `getProviderID` needs to make sure that those prerequisites are met. It feels okay because all those helper functions in this file are local and no one else uses them. It feels wrong because they can very easily become generic utility functions used somewhere else and break stuff if someone abuses them. Let's hear people out, it's hard for me to have a strong opinion about any possible approach here